src/swarm-lang/Swarm/Language/LSP/Definition.hs

```diff
@@ -0,0 +1,105 @@
+-- LSP support for finding and jumping to definitions
+-- SPDX-License-Identifier: BSD-3-Clause
+module Swarm.Language.LSP.Definition (
+  findDefinition,
+  DefinitionResult (..),
+)
+where
+
+import Data.List.NonEmpty qualified as NE
+import Data.Maybe (mapMaybe, maybeToList)
+import Data.Text.Utf16.Rope.Mixed qualified as R
+import Language.LSP.Protocol.Types qualified as J
+import Language.LSP.Protocol.Types qualified as LSP
+import Language.LSP.VFS (VirtualFile (VirtualFile), virtualFileText)
+import Swarm.Language.LSP.Position qualified as P
+import Swarm.Language.Parser (readTerm')
+import Swarm.Language.Parser.Core (ParserError, defaultParserConfig)
+import Swarm.Language.Pipeline (processParsedTerm)
+import Swarm.Language.Syntax (Located (..), SrcLoc, Syntax, Syntax' (Syntax'), Term' (..), Var)
+import Swarm.Language.TDVar (tdVarName)
+import Swarm.Language.Typecheck (ContextualTypeErr)
+
+data DefinitionResult = TError ContextualTypeErr | PError ParserError | Unsupported | NotFound | Found [J.Range]
+
+findDefinition ::
+  J.NormalizedUri ->
+  J.Position ->
+  VirtualFile ->
+  DefinitionResult
+findDefinition _ p vf@(VirtualFile _ _ myRope) =
+  either
+    PError
+    (maybe Unsupported findDef)
+    (readTerm' defaultParserConfig content)
+ where
+  content = virtualFileText vf
+  absolutePos =
+    R.charLength . fst $ R.charSplitAtPosition (P.lspToRopePosition p) myRope
+
+  -- build a list from the syntax tree and starting from the position of the cursor (the bottom).
+  -- search for the matching definition.
+  findDef :: Syntax -> DefinitionResult
+  findDef stx =
+    case processParsedTerm stx of
+      Left e -> TError e
+      Right pt -> do
+        let path = P.pathToPosition pt $ fromIntegral absolutePos
+        let usage = usageName $ NE.last path
+
+        case usage of
+          Nothing -> Unsupported
+          Just u -> do
+            let pathTerms = concatMap syntaxVars (NE.drop 1 . NE.reverse $ path)
+            case mapMaybe (maybeDefPosition u) pathTerms of
+              [] -> NotFound
+              ranges -> Found ranges
+
+  -- take a syntax element that we want to find the defintion for and
+  -- a possible syntax element that contains it's defintion
+  -- if this is the matching definition return the position
+  maybeDefPosition :: Var -> (SrcLoc, Var) -> Maybe LSP.Range
+  maybeDefPosition name (pos, name')
+    | name == name' = P.posToRange myRope pos
+    | otherwise = Nothing
+
+-- | find the name of the syntax element if it is a value level variable
+usageName :: Syntax' a -> Maybe Var
+usageName (Syntax' _ t _ _) = case t of
+  (TVar n) -> Just n
+  _ -> Nothing
+
+syntaxVars :: Syntax' a -> [(SrcLoc, Var)]
```

Comment:
I think maybe `syntaxVars` ought to have a return type of `Maybe (SrcLoc, Var)` rather than a list.  Any syntax node binds at most one variable.